### PR TITLE
[MM-25749] Implement non-cached autocomplete for mobile

### PR DESCRIPTION
**Pull request by iomodo**
#### Summary
This PR will add support of the slash command autocomplete on mobile, without caching the commands on client.

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-25749
Fixes https://mattermost.atlassian.net/browse/MM-27098

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes

#### Device Information
This PR was tested on: iphone 11, OS 13.5(simulator)

#### Screenshots
Simulator Screen Shot - iPhone 11 - 2020-06-26 at 00 17 09

### CODE REVIEW

File: app/components/autocomplete/slash_suggestion/non_cached/slash_suggestion.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+} from 'react-native';
+
+import AutocompleteDivider from 'app/components/autocomplete/autocomplete_divider';
+import {makeStyleSheetFromTheme} from 'app/utils/theme';
+
+import SlashSuggestionItem from '../slash_suggestion_item';
+import {Client4} from '@mm-redux/client';
+import {COMMANDS_TO_HIDE_ON_MOBILE} from '../cached/index';
+
+export default class SlashSuggestionNonCached extends PureComponent {
```
Comment (iomodo) [inline app/components/autocomplete/slash_suggestion/non_cached/slash_suggestion.js:18]: Note that I did not create a parent class here therefore `SlashSuggestionNonCached` and `SlashSuggestionCached` have some repeated code. Composition will not work, not sure if we should use inheritance.

File: app/components/autocomplete/slash_suggestion/slash_suggestion.js [hunk 2]
```diff
@@ -57,47 +59,71 @@ export default class SlashSuggestion extends PureComponent {
...
-            this.props.actions.getAutocompleteCommands(nextProps.currentTeamId);
+            const matches = this.filterSlashSuggestions(nextValue.substring(1), nextCommands);
+            this.updateSuggestions(matches);
+        } else if (isMinimumServerVersion(Client4.getServerVersion(), 5, 24)) {
+            const args = {
+                channel_id: this.props.channelId,
+                ...(this.props.rootId && {root_id: this.props.rootId, parent_id: this.props.rootId}),
+            };
+
+            Client4.getCommandAutocompleteSuggestionsList(nextProps.value, nextProps.currentTeamId, args).then(
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/slash_suggestion.js:94]: Why client calls in the component? this should be an action, then you  can use the result returned by the action no problem...

Also you should handle the case when the request fails.

File: app/components/autocomplete/slash_suggestion/slash_suggestion.js [hunk 3]
```diff
@@ -128,21 +166,23 @@ export default class SlashSuggestion extends PureComponent {
             });
         }
 
-        this.setState({
-            active: false,
-            suggestionComplete: true,
-        });
+        if (!isMinimumServerVersion(Client4.getServerVersion(), 5, 24)) {
```
Comment (enahum) [inline app/components/autocomplete/slash_suggestion/slash_suggestion.js:169]:
```suggestion
        if (!isMinimumServerVersion(Client4.getServerVersion(), 5, 26)) {
```

File: app/screens/edit_post/edit_post.js [hunk 4]
```diff
@@ -279,6 +279,8 @@ export default class EditPost extends PureComponent {
                         value={message}
                         nestedScrollEnabled={true}
                         onVisible={this.onAutocompleteVisible}
+                        rootId={this.props.post.root_id}
```
Comment (enahum) [inline app/screens/edit_post/edit_post.js:282]: Why do we want slash command autocomplete in edit post ?

File: app/components/autocomplete/slash_suggestion/slash_suggestion.js [hunk 5]
```diff
@@ -57,47 +59,71 @@ export default class SlashSuggestion extends PureComponent {
...
+                this.setState({
+                    lastCommandRequest: Date.now(),
+                });
+            }
 
-        if ((!nextCommands.length || dataIsStale)) {
-            this.props.actions.getAutocompleteCommands(nextProps.currentTeamId);
+            const matches = this.filterSlashSuggestions(nextValue.substring(1), nextCommands);
+            this.updateSuggestions(matches);
+        } else if (isMinimumServerVersion(Client4.getServerVersion(), 5, 24)) {
```
[the remark this answers is not in the document]
Comment (iomodo) [reply app/components/autocomplete/slash_suggestion/slash_suggestion.js:88]: Nope 24 will work fine. No changes needed on the server.

File: app/screens/edit_post/edit_post.js (hunk 4 above)
Comment (iomodo) [reply app/screens/edit_post/edit_post.js:282]: It was already there. When you edit post you can use command autocomplete. I thought it's a feature, isn't it? :D

Comment (enahum) [reply app/screens/edit_post/edit_post.js:282]: lol... it seems weird to me and I doubt it works,.. maybe create another ticket to remove slash commands from edit post ?

Comment (iomodo) [reply app/screens/edit_post/edit_post.js:282]: It's not executing for sure, but you can type and autocomplete commands. Should I push those changes here as well?

Comment (enahum) [reply app/screens/edit_post/edit_post.js:282]: If you can do it, that'll be great.. if not a new ticket is fine.

**Comment by enahum** [comment]
Thanks @DHaussermann I think we are going to go with it and add some telemetry to see how much the feature is being used with the known issue that is very bad UX on poor networks. Then based on the telemetry we can decide how to move fwd.

cc: @esethna @levb

File: app/screens/edit_post/edit_post.js (hunk 4 above)
Comment (iomodo) [reply app/screens/edit_post/edit_post.js]: @enahum FYI removed autocomplete on edit_post, as well as on edit_channel_info

**Comment by enahum** [comment]
@DHaussermann this should also be tested against the esr and expect no regressions

**Comment by DHaussermann** [comment]
Thanks @enahum  fir the reminder. Reverted labels for now. Will test on ESR.

**Comment by DHaussermann** [comment]
Tested 5.19.3 and 5.25.0 to cover ESR releases. (not sure when 5.25.0 will become the official ESR)

On those releases I found no impact when running this mobile build.

**Comment by amyblais** [comment]
@asaadmahmood This one needs UX review.

**Comment by enahum** [comment]
@iomodo please clear with @amyblais and UX if this is good to merge and cherry-pick as the deadline was on Friday :S

**Comment by amyblais** [comment]
Discussed with PM (Aaron) and apparently this is important for this release and they don't want to see this getting bumped.

@iomodo Can you help describe the level of risk this PR may cause? @DHaussermann From a QA point of view are we confident adding this to v1.34 or would it be better to push this to next release?

**Comment by DHaussermann** [comment]
From a QA perspective, the performance was deemed acceptable (see above) and a story exists to review future UX improvements. 

In fairness, it did miss the deadline because I did more regression testing on previous servers after Friday. I would have no objection to seeing this merged.

**Comment by iomodo** [comment]
@amyblais there is no change for servers later than 5.24. For newer ones there will be autocomplete fetching suggestion from server on every keystroke after the initial command. So I do not see big risk here.

**Comment by levb** [comment]
@amyblais Also to clarify - the new functionality does not apply to the existing commands, and Dylan tested that. We will be testing command-specific autocomplete implementations as we roll them out, separately; and have the option to delay those if bugs are discovered.

**Comment by amyblais** [comment]
@iomodo Based on comments from QA and dev, I think this is ok for to cherry-pick.

I'll work with Aaron and Lev to improve the process for getting features merged on time by the deadline in the future.

**Pull request merged**